### PR TITLE
Permute Distributed: migrate from AllToAll to Send/Recv

### PR DESCRIPTION
**Pull request by albestro**
Optimizations:
- Split unpacking of permutations from "all-at-once" to "local + others".
"locals" do not depend anymore on communication and can happen in parallel with the communication of "others".
- Migrate from MPI_AllToAllv to MPI_Send/MPI_Recv
  - Might reduce amount of data to be communicated (local are not communicated anymore)
  - Do not require anymore contiguousness of the MPI datatype
  For permute-row, this means not having anymore to transpose data, which implied a copy on a support memory (alternatively mat_in and mat_out...).
  - Not having to use `mat_in` as support memory, it was possible to fix const-correctness of the API

Additional Changelog:
- Test has been improved: now it also checks out-of-the-range invariability + there is a test-config for checking an edge case of the new implementation (related to local unpacking)
- Fix const-correctness also in `test_permutations_local`
- minor: Fix warning in MPI call inside gtest_listener
- some renamings to make some parts of the code more self-explanatory

_Note: I'm not super-happy about the code duplication introduced by `applyPermutationsFiltered` (more or less a copy of `applyPermutations`). I don't want to spend to much now on it, but in the near future I might have to do some other change on that part and your suggestion might be helpful._

### CODE REVIEW

File: include/dlaf/permutations/general/impl.h [hunk 1]
```diff
@@ -191,119 +236,135 @@ auto whenAllReadOnlyTilesArray(Matrix<const T, D>& matrix) {
...
+                ex::just() | dlaf::comm::internal::transformMPI([=](MPI_Request* req) {
+                  const SizeType nperms = send_counts[rank_partner_index];
+                  auto message = dlaf::comm::make_message(
+                      DataDescriptor<const T>(send_ptr +
+                                                  send_displs[rank_partner_index] * send_perm_stride,
+                                              C == Coord::Col ? nperms : vec_size,
+                                              C == Coord::Col ? vec_size : nperms, send_ld));
+
+                  DLAF_MPI_CHECK_ERROR(MPI_Isend(message.data(), message.count(), message.mpi_type(),
+                                                 rank_partner, 0, comm, req));
```
Comment (albestro) [inline include/dlaf/permutations/general/impl.h:288]: note-to-self: use a better TAG, in order to not rely on other constraints for avoiding mismatched communication.

@rasolca suggests to use `index_split` (wrt tridiagonal solver)

File: test/src/gtest_mpi_listener.cpp [hunk 2]
```diff
@@ -160,7 +160,8 @@ void MPIListener::OnTestEndAllRanks(const ::testing::TestInfo& test_info) const
 
 namespace internal {
 void mpi_send_string(const std::string& message, int to_rank) {
-  MPI_Send(message.c_str(), static_cast<int>(message.size()) + 1, MPI_CHAR, to_rank, 0, MPI_COMM_WORLD);
+  MPI_Send(const_cast<char*>(message.c_str()), static_cast<int>(message.size()) + 1, MPI_CHAR, to_rank,
+           0, MPI_COMM_WORLD);
```
Comment (msimberg) [inline test/src/gtest_mpi_listener.cpp:164]: FYI, also changed in #834.

File: include/dlaf/permutations/general/impl.h [hunk 3]
```diff
@@ -126,9 +129,51 @@ void applyPermutations(GlobalElementIndex out_begin, GlobalElementSize sz, SizeT
   }
 }
 
+template <class T, Device D, Coord C>
+void applyPermutationsFiltered(
+    GlobalElementIndex out_begin, GlobalElementSize sz, SizeType in_offset,
+    const matrix::Distribution& subm_dist, const SizeType* perm_arr,
+    const std::vector<pika::shared_future<matrix::Tile<const T, D>>>& in_tiles_fut,
+    const std::vector<matrix::Tile<T, D>>& out_tiles, std::function<bool(SizeType)> filter) {
```
Comment (msimberg) [inline include/dlaf/permutations/general/impl.h:137]: Regarding the duplication here, you could do:
```suggestion
    const std::vector<matrix::Tile<T, D>>& out_tiles, std::function<bool(SizeType)> filter = [](SizeType){ return false; }) {
```
If you also template the callable the compiler might completely optimize it away. The downside is another template parameter, and if you want to check that the callable is a `bool(SizeType)` the signature becomes more verbose.

Comment (msimberg) [inline include/dlaf/permutations/general/impl.h:137]: I realize the direction is arbitrary, so I don't feel too strongly about this, except that it should written in the documentation explicitly whether the predicate filters in or out.

My suggestion would be to rename `filter` to `pred` and have it filter items out when the predicate returns `false`. This would be consistent with https://en.cppreference.com/w/cpp/ranges/filter_view. However, I can see why the current option might be more intuitive as well, which is why the first part of this comment is the important one to avoid confusion.

File: include/dlaf/permutations/general/impl.h [hunk 4]
```diff
@@ -126,9 +129,51 @@ void applyPermutations(GlobalElementIndex out_begin, GlobalElementSize sz, SizeT
...
+  constexpr auto OC = orthogonal(C);
+  std::vector<SizeType> splits =
+      dlaf::util::interleaveSplits(sz.get<OC>(), subm_dist.blockSize().get<OC>(),
+                                   subm_dist.distanceToAdjacentTile<OC>(in_offset),
+                                   subm_dist.distanceToAdjacentTile<OC>(out_begin.get<OC>()));
+
+  const SizeType nperms = subm_dist.size().get<C>();
+
+  // Parallelized over the number of permutations
+  pika::for_loop(pika::execution::par, to_sizet(0), to_sizet(nperms), [&](SizeType i_perm) {
```
Comment (msimberg) [inline include/dlaf/permutations/general/impl.h:147]: This is for a follow-up, i.e. no change requested here, but I want to comment to spread awareness: calls to pika's parallel algorithms should be replaced with the asynchronous versions. In this particular case we could simply use `bulk`.

File: include/dlaf/permutations/general/impl.h (hunk 3 above)
Comment (albestro) [reply include/dlaf/permutations/general/impl.h:137]: It's a super good point.

I thought it was a good idea to check what other public APIs do...I looked at other functional approaches (e.g. [haskell](https://hackage.haskell.org/package/base-4.18.0.0/docs/Prelude.html#v:filter), [python](https://docs.python.org/3/library/functions.html?highlight=filter#filter)) and then I forgot to do apply the change 😛 

I fully agree with you to filter out when false. I'll do it.

File: test/unit/permutations/test_permutations_distributed.cpp [hunk 5]
```diff
@@ -33,86 +41,136 @@ struct PermutationsDistTestMC : public TestWithCommGrids {};
...
-  Matrix<T, Device::CPU> mat_out_h(dist);
+  const Distribution dist(size, block_size, grid.size(), grid.rank(), src_rank_index);
 
-  SizeType index_start = dist.globalElementFromGlobalTileAndTileElement<C>(i_begin, 0);
-  SizeType index_finish = dist.globalElementFromGlobalTileAndTileElement<C>(i_end, 0) +
-                          dist.tileSize(GlobalTileIndex(i_end, i_end)).get<C>();
-  dlaf::matrix::util::set(perms_h, [index_start, index_finish](GlobalElementIndex i) {
-    if (index_start > i.row() || i.row() >= index_finish)
-      return SizeType(0);
+  const auto [index_start, index_end] = tileToElementRange(n, nb, i_begin, i_last + 1);
```
Comment (albestro) [inline test/unit/permutations/test_permutations_distributed.cpp:107]: note: clang@12 complains because structured binding does not create variables but just "names", and it conflicts with what can be captured in lambdas (just variables).

**Comment by rasolca** [comment]
cscs-ci run

**Review by msimberg — APPROVED** [review]
